libcudacxx/include/cuda/std/__random/philox_engine.h

```diff
@@ -0,0 +1,575 @@
+        __os << __os.widen(' ');
+      }
+    }
+    __os << __os.widen(' ');
+
+    // output output buffer (__y_)
+    for (::cuda::std::size_t __i = 0; __i < word_count; ++__i)
+    {
+      __os << __e.__y_[__i];
+      if (__i < word_count - 1)
+      {
+        __os << __os.widen(' ');
+      }
+    }
+    __os << __os.widen(' ');
+
+    // output current position
+    __os << __e.__j_;
+
+    // restore flags & fill character
+    __os.flags(__flags);
+    __os.fill(__fill);
+
+    return __os;
+  }
+
+  //! This function streams a philox_engine in from a std::basic_istream.
+  //! @param is The basic_istream to stream from.
+  //! @param e The philox_engine to stream in.
+  //! @return is
+  template <typename _CharT, typename _Traits>
+  _CCCL_API friend ::std::basic_istream<_CharT, _Traits>&
+  operator>>(::std::basic_istream<_CharT, _Traits>& __is, philox_engine& __e)
+  {
+    using istream_type = ::std::basic_istream<_CharT, _Traits>;
+    using ios_base     = typename istream_type::ios_base;
+
+    // save old flags
+    const typename ios_base::fmtflags __flags = __is.flags();
+
+    __is.flags(ios_base::dec);
+
+    // input counter array (__x_)
+    for (::cuda::std::size_t __i = 0; __i < word_count; ++__i)
+    {
+      __is >> __e.__x_[__i];
+    }
+
+    // input key array (__k_)
+    for (::cuda::std::size_t __i = 0; __i < word_count / 2; ++__i)
+    {
+      __is >> __e.__k_[__i];
+    }
+
+    // input output buffer (__y_)
+    for (::cuda::std::size_t __i = 0; __i < word_count; ++__i)
+    {
+      __is >> __e.__y_[__i];
+    }
+
+    // input current position
+    __is >> __e.__j_;
+
+    // restore flags
+    __is.flags(__flags);
+
+    return __is;
+  }
+#endif // !_CCCL_COMPILER(NVRTC)
+
+private:
+  _CCCL_API constexpr void __increment_counter() noexcept
+  {
+    // Increment the big integer __x_ by 1, handling overflow.
+    _CCCL_PRAGMA_UNROLL_FULL()
+    for (::cuda::std::size_t __i = 0; __i < word_count; ++__i)
+    {
+      __x_[__i] = (__x_[__i] + 1) & max();
+      if (__x_[__i] != 0)
+      {
+        break;
+      }
+    }
+  }
+
+  [[nodiscard]] _CCCL_API constexpr auto __mulhilo_fallback(result_type __a, result_type __b) const noexcept
+  {
+    // Generic slow implementation
+    constexpr result_type __w_half  = word_size / 2;
+    constexpr result_type __lo_mask = (((result_type) 1) << __w_half) - 1;
+
+    result_type __lo  = __a * __b;
+    result_type __ahi = __a >> __w_half;
+    result_type __alo = __a & __lo_mask;
+    result_type __bhi = __b >> __w_half;
+    result_type __blo = __b & __lo_mask;
+
+    result_type __ahbl = __ahi * __blo;
+    result_type __albh = __alo * __bhi;
+
+    result_type __ahbl_albh = ((__ahbl & __lo_mask) + (__albh & __lo_mask));
+    result_type __hi        = __ahi * __bhi + (__ahbl >> __w_half) + (__albh >> __w_half);
+    __hi += __ahbl_albh >> __w_half;
+    __hi += ((__lo >> __w_half) < (__ahbl_albh & __lo_mask));
+
+    return ::cuda::std::make_pair(__hi, __lo);
```

Comment:
Nitpick: We are C++17 only, so we can skip `make_pair`
```suggestion
    return ::cuda::std::pair{__hi, __lo};
```